### DISCUSSION TITLE
Media controls on taskbar preview thumbnail (Windows)

### DISCUSSION THREAD

**Discussion by instantepiphany** [Taiko2k/Tauon#1019] [Ideas]
Windows (7 and later, I believe) have support for adding media controls to the taskbar preview thumbnail. 

![image](https://github.com/Taiko2k/TauonMusicBox/assets/5311587/fff244dd-5098-4398-9476-93a1a71c1e58)

This provides a very quick way for the user to adjust playback without opening the main interface. Most media apps on windows have these buttons.

The documentation is here: https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-itaskbarlist3-thumbbaraddbuttons

I'd enjoy making an attempt at this. Would you be open to this feature being added (windows only) assuming my PR was acceptable quality?

**Comment by Taiko2k** [comment]
Certainly this would we a fine feature. Just how easy it is to implement is another thing.

If you wanted to get started on this, a good place to start would be the file [TauonSMTC](https://github.com/Taiko2k/TauonMusicBox/tree/master/src/smtc/TauonSMTC)/dllmain.cpp, which already implements the Windows SMTC interface. Theoretically it shouldn't be hard to extend that code to support the taskbar preview... 

Except for an issue I ran into in that I couldn't seem to find a way to pass the HWND from SDL to Windows through my external DLL. Windows would always seem to reject it. I never found a solution, so my hacky workaround was to create a fake dummy Window and use its HWND instead. But for this taskbar feature we would surely need the real HWND of Tauon's window. I imagine the issue is something like SDL creating the window on its own thread and Windows refuses to use the HWND from different thread, something like that, or maybe I was type converting the value wrong, I'm not sure. Hard problem. 😕

**Reply by Taiko2k** [reply]
If my above assumption is true it may be that its impossible to do this without patching the SDL source code directly, if so I think that's beyond the scope for this project.

**Reply by instantepiphany** [reply]
Ouch, that is an annoying limitation! I'll give it a try but doubt I'll be able to figure out that where you couldn't - I haven't touched SDL for years. So hopefully there is some other workaround, or it doesn't require HWND.
